ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,491 @@
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// amt.set(2, "two".to_owned());
+    /// amt.set(1, "one".to_owned());
+    ///
+    /// let vals = amt.values().collect::<Vec<_>>();
+    /// assert_eq!(vals, ["one", "two"]);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn values(&self) -> impl Iterator<Item = &V> {
+        self.iter().map(|res| {
+            res.map(|(_, v)| v)
+                .expect("Failed to generate iterator from AMT values")
+        })
+    }
+
+    /// Generates an iterator containing the keys of the AMT, in sorted order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// amt.set(2, "two".to_owned());
+    /// amt.set(1, "one".to_owned());
+    ///
+    /// let keys = amt.keys().collect::<Vec<_>>();
+    /// assert_eq!(keys, [1, 2]);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn keys(&self) -> impl Iterator<Item = usize> + '_ {
+        self.iter().map(|res| {
+            res.map(|(k, _)| k)
+                .expect("Failed to generate iterator from AMT keys")
+        })
+    }
+
+    /// Iterate over the AMT starting at the given key. This can be used to implement "ranged"
+    /// iteration:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// // Create an AMT with 5 keys.
+    /// let mut amt = Amt::new(store);
+    /// let kvs: Vec<usize> = (0..=5).collect();
+    ///
+    /// let _ = kvs
+    ///     .iter()
+    ///     .map(|k| amt.set(u64::try_from(*k).unwrap(), k.to_string()))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// let kvs = kvs
+    ///     .iter()
+    ///     .map(|k| (k.clone(), k.to_string()))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// // Read 2 elements.
+    /// let mut results = amt.iter().take(2).collect::<Result<Vec<_>, _>>().unwrap();
+    /// assert_eq!(results.len(), 2);
+    ///
+    /// // Read the rest.
+    /// for res in amt.iter_from(results.last().unwrap().0 + 1) {
+    ///     results.push(res.unwrap());
+    /// }
+    ///
+    /// // Assert that we got out what we put in.
+    /// let results: Vec<_> = results
+    ///     .into_iter()
+    ///     .map(|(k, v)| (k.clone(), v.clone()))
+    ///     .collect();
+    ///
+    /// assert_eq!(kvs, results);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn iter_from(&self, key: usize) -> Vec<Result<(usize, &V), crate::Error>> {
+        let mut results = Vec::new();
+        for res in self.iter() {
+            let (k, v) = res.expect("Failed to generate iterator from AMT");
+            if k >= key {
```

Comment:
This needs to seek to the correct position without iterating (see `for_each_ranged`).